### PR TITLE
HB-05

### PR DESCRIPTION
**Pull request by KostiantynVorobiov**

### CODE REVIEW

File: src/main/java/com/cinema/security/AuthenticationServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        }
+        throw new AuthenticationException("Incorrect login or password");
+    }
+
+    @Override
+    public User register(String email, String password) {
+        User user = new User();
+        user.setEmail(email);
+        user.setPassword(password);
+        userService.add(user);
```
Comment (vremere-a) [inline src/main/java/com/cinema/security/AuthenticationServiceImpl.java]: return userService.add(user);

**Review by vremere-a — APPROVED** [review]
well done

File: src/main/java/com/cinema/dao/impl/UserDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @Override
+    public Optional<User> findByEmail(String email) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<User> getUserByEmailQuery = session.createQuery("FROM User WHERE email = :email",
+                    User.class);
+            getUserByEmailQuery.setParameter("email", email);
+            User user = getUserByEmailQuery.uniqueResult();
+            return Optional.ofNullable(user);
```
Comment (Sofasmile) [inline src/main/java/com/cinema/dao/impl/UserDaoImpl.java]:
```suggestion
            return query.uniqueResultOptional();
```

File: src/main/java/com/cinema/security/AuthenticationServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+public class AuthenticationServiceImpl implements AuthenticationService {
+
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> userFromDb = userService.findByEmail(email);
+        if (userFromDb.isPresent() && userFromDb.get().getPassword().equals(HashUtil
+                . hashPassword(password, userFromDb.get().getSalt()))) {
```
Comment (Sofasmile) [inline src/main/java/com/cinema/security/AuthenticationServiceImpl.java]:
```suggestion
        if (userFromDb.isPresent() && userFromDb.get().getPassword()
                .equals(HashUtil.hashPassword(password, userFromDb.get().getSalt()))) {
```

File: src/main/java/com/cinema/util/HashUtil.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        StringBuilder hashedPassword = new StringBuilder();
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance(ALGORITHM);
+            byte[] digest = messageDigest.digest(password.getBytes());
+            messageDigest.update(salt);
+            for (byte b : digest) {
+                hashedPassword.append(String.format("%02x", b));
+            }
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("Can't get hash from password");
```
Comment (Sofasmile) [inline src/main/java/com/cinema/util/HashUtil.java]:
```suggestion
            throw new RuntimeException("Can't get hash from password", e);
```

File: src/main/java/com/cinema/security/AuthenticationServiceImpl.java (hunk 1 above)
Comment (KostiantynVorobiov) [reply src/main/java/com/cinema/security/AuthenticationServiceImpl.java]: Done

File: src/main/java/com/cinema/dao/impl/UserDaoImpl.java (hunk 2 above)
Comment (KostiantynVorobiov) [reply src/main/java/com/cinema/dao/impl/UserDaoImpl.java]: Done

File: src/main/java/com/cinema/security/AuthenticationServiceImpl.java (hunk 3 above)
Comment (KostiantynVorobiov) [reply src/main/java/com/cinema/security/AuthenticationServiceImpl.java]: Done

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**